### PR TITLE
Sb 8241

### PR DESCRIPTION
**Pull request by amolnin**

### CODE REVIEW

File: service/app/controllers/usermanagement/UserCrudController.java [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+        request().body().asJson(),
+        (req) -> {
+          Request request = (Request) req;
+          UserRequestValidator.validateCreateUserV2(request);
+          HashMap<String, Object> innerMap = new HashMap<>();
+          innerMap.put(JsonKey.USER, request.getRequest());
+          request.setRequest(innerMap);
+          return null;
+        },
+        JsonKey.VERSION_2,
```
Comment (bvinayakumar) [inline service/app/controllers/usermanagement/UserCrudController.java]: It is confusing to pass JsonKey.VERSION_2 and JsonKey.VERSION as they do not correspond to pathId and pathVariable. Rather explicitly set it in the lambda function within the request.

File: service/app/controllers/usermanagement/UserCrudController.java [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+        (req) -> {
+          Request request = (Request) req;
+          String accessToken =
+              request().getHeader(HeaderParam.X_Authenticated_User_Token.getName());
+          request.getRequest().put(HeaderParam.X_Authenticated_User_Token.getName(), accessToken);
+          UserRequestValidator.validateUpdateUser(request);
+          if (null != ctx().flash().get(JsonKey.IS_AUTH_REQ)
+              && Boolean.parseBoolean(ctx().flash().get(JsonKey.IS_AUTH_REQ))) {
+            validateAuthenticity(request);
+            HashMap<String, Object> innerMap = new HashMap<>();
```
Comment (bvinayakumar) [inline service/app/controllers/usermanagement/UserCrudController.java]: No need to use an inner map. Change MW accordingly.

File: service/app/controllers/usermanagement/UserCrudController.java [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+
+    return handleRequest(
+        ActorOperations.UPDATE_USER.getValue(),
+        request().body().asJson(),
+        (req) -> {
+          Request request = (Request) req;
+          String accessToken =
+              request().getHeader(HeaderParam.X_Authenticated_User_Token.getName());
+          request.getRequest().put(HeaderParam.X_Authenticated_User_Token.getName(), accessToken);
+          UserRequestValidator.validateUpdateUser(request);
```
Comment (bvinayakumar) [inline service/app/controllers/usermanagement/UserCrudController.java]: validateUpdateUser -> validateUpdateUserRequest

File: service/app/controllers/usermanagement/UserCrudController.java [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+          if (null != ctx().flash().get(JsonKey.IS_AUTH_REQ)
+              && Boolean.parseBoolean(ctx().flash().get(JsonKey.IS_AUTH_REQ))) {
+            validateAuthenticity(request);
```
Comment (bvinayakumar) [inline service/app/controllers/usermanagement/UserCrudController.java]: Move the auth validation as well to validateUpdateUserRequest.

File: service/app/controllers/usermanagement/UserCrudController.java [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+   */
+  public Promise<Result> getUserProfile(String userId) {
+    final String requestedFields = request().getQueryString(JsonKey.FIELDS);
+
+    return handleRequest(
+        ActorOperations.GET_PROFILE.getValue(),
+        null,
+        (req) -> {
+          Request request = (Request) req;
+          HashMap<String, Object> innerMap = new HashMap<>();
```
Comment (bvinayakumar) [inline service/app/controllers/usermanagement/UserCrudController.java]: No need to use an inner map. Change MW accordingly.

File: service/app/controllers/usermanagement/UserCrudController.java [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+  public Promise<Result> getUserDetailsByLoginId() {
+
+    final String requestedFields = request().getQueryString(JsonKey.FIELDS);
+    return handleRequest(
+        ActorOperations.GET_USER_DETAILS_BY_LOGINID.getValue(),
+        request().body().asJson(),
+        (req) -> {
+          Request request = (Request) req;
+          UserRequestValidator.validateVerifyUser(request);
+          HashMap<String, Object> innerMap = new HashMap<>();
```
Comment (bvinayakumar) [inline service/app/controllers/usermanagement/UserCrudController.java]: No need to use an inner map. Change MW accordingly.

File: service/app/controllers/usermanagement/UserCrudController.java [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+   * This method will do the user search for Elastic search. this will internally call composite
+   * search api.
+   *
+   * @return Promise<Result>
+   */
+  public Promise<Result> search() {
+    return handleSearchRequest(
+        ActorOperations.COMPOSITE_SEARCH.getValue(),
+        request().body().asJson(),
+        orgRequest -> {
```
Comment (bvinayakumar) [inline service/app/controllers/usermanagement/UserCrudController.java]: orgRequest -> request (or) searchRequest (or) userSearchRequest

File: service/app/controllers/usermanagement/UserCrudController.java [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+    final String requestedFields = request().getQueryString(JsonKey.FIELDS);
+    return handleRequest(
+        ActorOperations.GET_USER_DETAILS_BY_LOGINID.getValue(),
+        request().body().asJson(),
+        (req) -> {
+          Request request = (Request) req;
+          UserRequestValidator.validateVerifyUser(request);
+          HashMap<String, Object> innerMap = new HashMap<>();
+          innerMap.put(JsonKey.USER, request.getRequest());
+          innerMap.put(JsonKey.FIELDS, requestedFields);
```
Comment (bvinayakumar) [inline service/app/controllers/usermanagement/UserCrudController.java]: Put the fields in the context.

File: service/app/controllers/usermanagement/UserCrudController.java [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+        null,
+        null,
+        getAllRequestHeaders(request()),
+        EsType.user.getTypeName());
+  }
+
+  private void validateAuthenticity(Request reqObj) {
+
+    if (ctx().flash().containsKey(JsonKey.AUTH_WITH_MASTER_KEY)) {
+      validateWithClient(reqObj);
```
Comment (bvinayakumar) [inline service/app/controllers/usermanagement/UserCrudController.java]: Let us drop the check on validateWithClient. It is not used.

File: service/app/controllers/usermanagement/UserProfileController.java [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+
+public class UserProfileController extends BaseController {
+
+  /**
+   * Get all the social media types supported
+   *
+   * @return
+   */
+  public Promise<Result> getMediaTypes() {
+    String userId =
```
Comment (bvinayakumar) [inline service/app/controllers/usermanagement/UserProfileController.java]: For getting media types, userId is unused.

File: service/app/controllers/usermanagement/UserController.java [hunk 11]
```diff
@@ -36,173 +30,6 @@
  */
 public class UserController extends BaseController {
```
Comment (bvinayakumar) [inline service/app/controllers/usermanagement/UserController.java]: Remove this file. Rename UserCrudController as UserController.

File: service/app/controllers/usermanagement/UserProfileController.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+  }
+
+  /**
+   * This method will add or update user profile visibility control. User can make all field as
+   * private except name. any private filed of user is not search-able.
+   *
+   * @return Promise<Result>
+   */
+  public Promise<Result> profileVisibility() {
+    if (null != ctx().flash().get(JsonKey.IS_AUTH_REQ)
```
[the remark this answers is not in the document]
Comment (bvinayakumar) [reply service/app/controllers/usermanagement/UserProfileController.java]: What is the resolution for this?

[the remark this answers is not in the document]
Comment (amolnin) [reply service/app/controllers/usermanagement/UserProfileController.java]: my bad, i checked the method initially. 
And forgot to apply the solution.

File: service/app/controllers/usermanagement/UserProfileController.java [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+    final boolean isAuthRequired =
+        null != ctx().flash().get(JsonKey.IS_AUTH_REQ)
+            && Boolean.parseBoolean(ctx().flash().get(JsonKey.IS_AUTH_REQ));
+    return handleRequest(
+        ActorOperations.PROFILE_VISIBILITY.getValue(),
+        request().body().asJson(),
+        (req) -> {
+          Request request = (Request) req;
+          if (isAuthRequired) {
+            BaseRequestValidator.validateUserId(request, JsonKey.USER_ID);
```
Comment (bvinayakumar) [inline service/app/controllers/usermanagement/UserProfileController.java]: Move this check to UserProfileRequestValidator

File: service/app/controllers/usermanagement/UserProfileController.java [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+          }
+          UserRequestValidator.validateProfileVisibility(request);
```
Comment (bvinayakumar) [inline service/app/controllers/usermanagement/UserProfileController.java]: Create UserProfileRequestValidator

**Pull request merged**